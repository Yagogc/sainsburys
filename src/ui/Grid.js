import styled from "styled-components";

export const GridContainer = styled.main`
  width: 100%;
  max-width: 800px;
  padding: 10px;
  margin: 0 auto;
  text-align: center;
  display: grid;
  grid-gap: 10px;
  grid-template-columns: repeat(auto-fill, minmax(150px, 1fr));
  align-items: center;
  justify-items: center;
`;

export const GridItem = styled.img`
  width: 100%;
  height: 150px;
  color: red;
  object-fit: cover;
  border: 2px solid ${props => (props.isSelected ? "red" : "#8897a2")}
  transform: scale3D(0.9, 0.9, 1);
  box-shadow: 0 1px 3px rgba(0, 0, 0, 0.12), 0 1px 2px rgba(0, 0, 0, 0.24);
  transition: all 0.3s cubic-bezier(0.25, 0.8, 0.25, 1);
  cursor: pointer;

  &:hover {
    transform: scale3D(1, 1, 1);
    box-shadow: 0 14px 28px rgba(0, 0, 0, 0.25), 0 10px 10px rgba(0, 0, 0, 0.22);
  }
`;
