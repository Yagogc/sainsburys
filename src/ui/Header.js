import styled from "styled-components";

export const Header = styled.header`
  background: #f90;
  padding: 10px;
  margin-bottom: 15px;
  text-align: center;
  box-shadow: 0 1px 3px rgba(0, 0, 0, 0.12), 0 1px 2px rgba(0, 0, 0, 0.24);
`;

export const HeaderTitle = styled.h1`
  color: #fff;
`;
