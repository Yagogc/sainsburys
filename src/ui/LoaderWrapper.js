import styled from "styled-components";

export const LoaderWrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  min-height: calc(100vh - 58px);
`;
