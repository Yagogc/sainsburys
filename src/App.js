import React, { Component } from "react";
import Loader from "react-loader-spinner";

import { GridContainer, GridItem } from "./ui/Grid";
import fetchJsonp from "fetch-jsonp";
import { Header, HeaderTitle } from "./ui/Header";
import { ErrorMsg } from "./ui/Error";
import formatId from "./utils/formatid.js";
import { LoaderWrapper } from "./ui/LoaderWrapper";

class App extends Component {
  state = {
    items: "",
    selectedItems: [],
    loading: true,
    error: false
  };

  componentDidMount() {
    const self = this;

    this.getSessionStorage();

    fetchJsonp(
      "https://api.flickr.com/services/feeds/photos_public.gne?tags=sainsburys&tagmode=all&format=json",
      {
        jsonpCallbackFunction: "jsonFlickrFeed"
      }
    )
      .then(function(response) {
        return response.json();
      })
      .then(function(json) {
        console.log(json);
        console.log(json.items);
        self.setState({
          items: json.items,
          loading: false
        });
      })
      .catch(function(ex) {
        console.log("parsing failed", ex);
        self.setState({
          loading: false,
          error: true
        });
      });
  }

  componentDidUpdate() {
    sessionStorage.setItem("selectedItems", this.state.selectedItems);
  }

  getSessionStorage() {
    if (sessionStorage["selectedItems"] !== undefined) {
      const itemsStored = sessionStorage.getItem("selectedItems").split(",");
      this.setState({
        selectedItems: [...itemsStored]
      });
    }
  }

  selectItem(item) {
    item = formatId(item);
    const result = this.state.selectedItems.find(
      selectedItem => selectedItem === item
    );
    if (result) {
      const newSelectedItems = this.state.selectedItems.filter(
        selectedItem => selectedItem !== item
      );
      this.setState({
        selectedItems: [...newSelectedItems]
      });
    } else {
      this.setState({
        selectedItems: [...this.state.selectedItems, item]
      });
    }
  }
  isSelected(item) {
    item = formatId(item);
    const result = this.state.selectedItems.find(
      selectedItem => selectedItem === item
    );
    // console.log("isSelected: ", result);
    // console.log("isSelected2: ", result ? true : false);
    return result ? true : false;
  }
  render() {
    const { items } = this.state;
    return (
      <React.Fragment>
        <Header>
          <HeaderTitle>Sainsbury's</HeaderTitle>
        </Header>
        {this.state.loading && (
          <LoaderWrapper>
            <Loader type="TailSpin" color="#f90" height="40" width="40" />
          </LoaderWrapper>
        )}
        {this.state.error && (
          <ErrorMsg>Flickr Feed error. Please reload the page.</ErrorMsg>
        )}
        <GridContainer>
          {items &&
            items.map((item, i) => {
              return (
                <GridItem
                  key={i}
                  src={item.media.m}
                  alt={item.title}
                  onClick={e => this.selectItem(item.link)}
                  isSelected={this.isSelected(item.link)}
                />
              );
            })}
        </GridContainer>
      </React.Fragment>
    );
  }
}

export default App;
