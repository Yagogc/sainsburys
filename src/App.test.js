import React from "react";
import { shallow, mount, render } from "enzyme";
import "mock-local-storage";
import App from "./App";
import { Header } from "./ui/Header";

describe("App", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<App />);
  });

  it("renders correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });
  it("should render a <header />", () => {
    expect(wrapper.find("header").length).toEqual(1);
  });

  it("should render the Header Component", () => {
    expect(wrapper.containsMatchingElement(<Header />)).toEqual(true);
  });
});
