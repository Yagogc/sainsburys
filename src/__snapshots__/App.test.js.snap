// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <styled.header>
          <styled.h1>
                    Sainsbury's
          </styled.h1>
</styled.header>,
        <styled.div>
          <Loader
                    color="#f90"
                    height="40"
                    type="TailSpin"
                    visible={true}
                    width="40"
          />
</styled.div>,
        false,
        <styled.main />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <styled.h1>
            Sainsbury's
</styled.h1>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Sainsbury's",
          },
          "ref": null,
          "rendered": "Sainsbury's",
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <Loader
            color="#f90"
            height="40"
            type="TailSpin"
            visible={true}
            width="40"
/>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "color": "#f90",
            "height": "40",
            "type": "TailSpin",
            "visible": true,
            "width": "40",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "",
        },
        "ref": null,
        "rendered": "",
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <styled.header>
            <styled.h1>
                        Sainsbury's
            </styled.h1>
</styled.header>,
          <styled.div>
            <Loader
                        color="#f90"
                        height="40"
                        type="TailSpin"
                        visible={true}
                        width="40"
            />
</styled.div>,
          false,
          <styled.main />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <styled.h1>
              Sainsbury's
</styled.h1>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Sainsbury's",
            },
            "ref": null,
            "rendered": "Sainsbury's",
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <Loader
              color="#f90"
              height="40"
              type="TailSpin"
              visible={true}
              width="40"
/>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "color": "#f90",
              "height": "40",
              "type": "TailSpin",
              "visible": true,
              "width": "40",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "",
          },
          "ref": null,
          "rendered": "",
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
